sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -65,35 +92,96 @@ class EquivalentExpressions {
+  //          false expressions in `If` may not get accessed, according to the predicate
+  //          expression. We should only recurse into the predicate expression.
+  //   3. CaseWhen: like `If`, the children of `CaseWhen` only get accessed in a certain
+  //                condition. We should only recurse into the first condition expression as it
+  //                will always get accessed.
+  //   4. Coalesce: it's also a conditional expression, we should only recurse into the first
+  //                children, because others may not get accessed.
+  private def childrenToRecurse(expr: Expression): Seq[Expression] = expr match {
+    case _: CodegenFallback => Nil
+    case i: If => i.predicate :: Nil
+    case c: CaseWhen => c.children.head :: Nil
+    case c: Coalesce => c.children.head :: Nil
+    case s: SortPrefix => s.child.child :: Nil
+    case other => other.children
+  }
+
+  /**
+   * Adds the expression to this data structure recursively. Stops if a matching expression
+   * is found. That is, if `expr` has already been added, its children are not added.
+   */
+  def addExprTree(
+      expr: Expression,
+      exprMap: EquivalenceMap = this.equivalenceMap): Unit = {
+    val skip = skipExpr(expr)
 
-    // There are some special expressions that we should not recurse into all of its children.
-    //   1. CodegenFallback: it's children will not be used to generate code (call eval() instead)
-    //   2. If: common subexpressions will always be evaluated at the beginning, but the true and
-    //          false expressions in `If` may not get accessed, according to the predicate
-    //          expression. We should only recurse into the predicate expression.
-    //   3. CaseWhen: like `If`, the children of `CaseWhen` only get accessed in a certain
-    //                condition. We should only recurse into the first condition expression as it
-    //                will always get accessed.
-    //   4. Coalesce: it's also a conditional expression, we should only recurse into the first
-    //                children, because others may not get accessed.
-    def childrenToRecurse: Seq[Expression] = expr match {
-      case _: CodegenFallback => Nil
-      case i: If => i.predicate :: Nil
-      case c: CaseWhen => c.children.head :: Nil
-      case c: Coalesce => c.children.head :: Nil
-      case other => other.children
+    if (!skip && !addExpr(expr, exprMap)) {
+      childrenToRecurse(expr).foreach(addExprTree(_, exprMap))
     }
+  }
+
+  /**
+   * Adds the expression to structural data structure recursively. Returns false if this doesn't add
+   * the input expression actually.
+   */
+  def addStructuralExprTree(expr: Expression): Boolean = {
+    val skip = skipExpr(expr) || expr.isInstanceOf[CodegenFallback]
 
-    if (!skip && !addExpr(expr)) {
-      childrenToRecurse.foreach(addExprTree)
+    if (!skip && addStructExpr(expr)) {
+      childrenToRecurse(expr).foreach(addStructuralExprTree)
```

Comment:
nit: do we want to add `(_)`?

Reply:
addExprTree doesn't add (_) too. Just followed it. If no special reason, I will leave it.

Reply:
I am neutral on this.   
I was curious why [this line](https://github.com/apache/spark/pull/25717/files#diff-8bcc5aea39c73d4bf38aef6f6951d42cR1042) added `(_)`, but here does not add.

Reply:
At that line, if not add a (_) to call addExprTree, will see compilation error:

```
[error]  found   : (org.apache.spark.sql.catalyst.expressions.Expression, equivalentExpressions.EquivalenceMap) => Unit                                                               
[error]     (which expands to)  (org.apache.spark.sql.catalyst.expressions.Expression, scala.collection.mutable.HashMap[equivalentExpressions.Expr,scala.collection.mutable.ArrayBuff$r[org.apache.spark.sql.catalyst.expressions.Expression]]) => Unit                                                                                                                     
[error]  required: org.apache.spark.sql.catalyst.expressions.Expression => ?               
[error]     expressions.foreach(equivalentExpressions.addExprTree)     
[error]                                               ^                                                                                                                               
```

Because addExprTree actually needs to arguments, it doesn't match foreach's argument type.